SuperChamps/contracts/Metagame/SCMetagamePool.sol

```diff
@@ -0,0 +1,135 @@
+        if(!success) {
+            revert UnableToTransferTokens(msg.sender, amount_);
+        }
+        _stake(msg.sender, amount_);
+    }
+
+    /// @notice Stake tokens for another address
+    /// @param staker_ The address to stake tokens for
+    /// @param amount_ Quantity of tokens to stake
+    /// @dev Tokens are transferred from msg.sender and credited to staker_
+    function stakeFor(address staker_, uint256 amount_) external {
+        bool success = token.transferFrom(msg.sender, address(this), amount_);
+        if(!success) {
+            revert UnableToTransferTokens(msg.sender, amount_);
+        }
+        _stake(staker_, amount_);
+    }
+
+    /// @notice Approve an address to spend staked tokens
+    /// @param spender_ The address to approve
+    /// @param amount_ Quantity of tokens to approve spending
+    /// @dev This mirrors approve(...) defined in ERC20 standard
+    function approve(address spender_, uint256 amount_) external {
+        _user_to_approved_spend[msg.sender][spender_] += amount_;
+    }
+
+    /// @notice Spend tokens from a stakers staked tokens
+    /// @param amount_ Quantity of tokens to unstake/spend
+    /// @param staker_ The address of the staker
+    /// @param receiver_ The address of the reciever of the spent tokens
+    /// @dev This does NOT trigger an Unstake event, allowing stakers to spend tokens without penalty
+    /// @dev This is expected to be called from the spending contract, which must be first permissioned through a direct call to approve(...)
+    function spend(uint256 amount_, address staker_, address receiver_) external {
+        if(staker_ != msg.sender) {
+            _user_to_approved_spend[staker_][msg.sender] -= amount_;
+        }
+
+        uint256 _balance = _unstake(amount_, msg.sender, receiver_);
+        emit SpendFromStake(staker_, amount_, _balance);
+    }
+
+    /// @notice Unstake tokens and transfer to staker
+    /// @param amount_ Quantity of tokens to unstake
+    /// @dev This triggers an Unstake event which can cause penalties in the metagame
+    function unstake(uint256 amount_) external {
+        uint256 _balance = _unstake(amount_, msg.sender, msg.sender);
+        emit Unstake(msg.sender, amount_, _balance);
+    }
+
+    /// @notice Returns the list of all of the users staking checkpoint timestamps
+    /// @param staker_ Address of the staker
+    function checkpoint_timestamps(address staker_) public view returns (uint256[] memory) {
+        return _user_checkpoints[staker_];
+    }
+
+    /// @notice Returns the list of all of the user's checkpoint balances from a list of timestamps
+    /// @param staker_ Address of the staker
+    /// @param checkpoint_timestamps_ A list of checkpoints timestamps, retrievable with checkpoint_timestamps(...)
+    function checkpoints(address staker_, uint256[] memory checkpoint_timestamps_) public view returns (uint256[] memory _balances_) {
```

Comment:
@figs999 can this function just take staker address as input and use checkpoint_timestamps(staker_) call to get the second param?

Reply:
No, because it would end up making it impossible to call this method eventually. The gas limit makes a hard cap on the number of iterations that can be done, even in a view.

The 2-step process makes it so that a reasonably sized subset of the checkpoints can be queried.